cmd/origin_token.go

```diff
@@ -0,0 +1,310 @@
+			// feels wrong to correct an explicitly-provided version that isn't correct,
+			// so in that event, fail.
+			if reqClaim == "ver" {
+				verPattern := `^scitokens:2\.[0-9]+$`
+				re := regexp.MustCompile(verPattern)
+
+				if !re.MatchString(val) {
+					errMsg := "The provided version '" + val +
+						"' is not valid. It must match 'scitokens:<version>', where version is of the form 2.x"
+					return errors.New(errMsg)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func verifyCreateWLCG(claimsMap *map[string]string) error {
+	/*
+		Don't check for the following claims because ALL base tokens have them:
+		- iat
+		- exp
+		- nbf
+		- iss
+		- jti
+	*/
+	if len(*claimsMap) == 0 {
+		return errors.New("To create a valid wlcg, the 'aud' and 'sub' claims must be passed, but none were found.")
+	}
+
+	requiredClaims := []string{"sub", "wlcg.ver", "aud"}
+	for _, reqClaim := range requiredClaims {
+		if val, exists := (*claimsMap)[reqClaim]; !exists {
+			// we can set wlcg.ver because we know what it should be
+			if reqClaim == "wlcg.ver" {
+				(*claimsMap)["wlcg.ver"] = "1.0"
+			} else {
+				// We can't set the rest
+				errMsg := "The claim '" + reqClaim +
+					"' is required for the wlcg profile, but it could not be found."
+				return errors.New(errMsg)
+			}
+		} else {
+			if reqClaim == "wlcg.ver" {
+				verPattern := `^1\.[0-9]+$`
+				re := regexp.MustCompile(verPattern)
+				if !re.MatchString(val) {
+					errMsg := "The provided version '" + val + "' is not valid. It must be of the form '1.x'"
+					return errors.New(errMsg)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func parseClaims(claims []string) (map[string]string, error) {
+	claimsMap := make(map[string]string)
+	for _, claim := range claims {
+		// Split by the first "=" delimiter
+		parts := strings.SplitN(claim, "=", 2)
+		if len(parts) < 2 {
+			errMsg := "The claim '" + claim + "' is invalid. Did you forget an '='?"
+			return nil, errors.New(errMsg)
+		}
+		key := parts[0]
+		val := parts[1]
+
+		if existingVal, exists := claimsMap[key]; exists {
+			claimsMap[key] = existingVal + " " + val
+		} else {
+			claimsMap[key] = val
+		}
+	}
+	return claimsMap, nil
+}
+
+func CreateEncodedToken(claimsMap map[string]string, profile string, lifetime int) (string, error) {
+	var err error
+	if profile != "" {
+		if profile == "scitokens2" {
+			err = verifyCreateSciTokens2(&claimsMap)
+			if err != nil {
+				return "", errors.Wrap(err, "Token does not conform to scitokens2 requirements")
+			}
+		} else if profile == "wlcg" {
+			err = verifyCreateWLCG(&claimsMap)
+			if err != nil {
+				return "", errors.Wrap(err, "Token does not conform to wlcg requirements")
+			}
+		} else {
+			errMsg := "The provided profile '" + profile +
+				"' is not recognized. Valid options are 'scitokens2' or 'wlcg'"
+			return "", errors.New(errMsg)
+		}
+	}
+
+	lifetimeDuration := time.Duration(lifetime)
+	// Create a jti using uuid4. This will be added to all tokens.
```

Comment:
It seems that UUIDs aren't built-in in golang; instead of adding a new dependency, let's do the following instead:

https://github.com/PelicanPlatform/pelican/pull/167/files#diff-9ecc873c07e2491d63b175c34ad6f290375f7932c2849c36cd7d106312953044R54-R58